### ISSUE TITLE
Client-requested resynchronization fires the UI detach listener on a live UI, destroying the scope of a tab that is still open

### ISSUE THREAD

**Issue by mvysny** [mvysny/vaadin-tab-scope#6]
## Summary

A client-requested **resynchronization** fires the UI's `DetachEvent` on a UI that is alive,
attached and not closing. `TabScope.init()`'s detach listener (`TabScope.java:402`) treats that as a
real detach, so a *live* tab loses its scope.

`ServerRpcHandler#handleUnloadBeaconRequest`'s sibling branch (Flow 25.2.1,
`ServerRpcHandler.java:455`) handles `rpcRequest.isResynchronize()` by calling:

```java
// Run detach listeners and re-attach all nodes again to the state tree, in order to send
// changes for a full re-build of the client-side state tree in the response
ui.getInternals().getStateTree().prepareForResync();
```

`StateNode#prepareForResync` (`StateNode.java:458`) then does exactly that — `fireDetachListeners()`
over the whole node tree, bottom-up, followed by re-fired attach listeners. The UI's own root node is
in that tree, so `ComponentMapping#onDetach` → `ComponentUtil.onComponentDetach(ui)` fires the UI's
`DetachEvent`. Nothing is actually detached: the UI stays in the session, and `isClosing()` is false.

Flow logs a WARN when this happens ("Resynchronizing UI by client's request…"), and the trigger is an
ordinary network fact — a lost UIDL message, or a proxy/load balancer terminating a push connection.

What a detach listener can observe, measured in both cases:

| event | `ui.isAttached()` | `ui.isClosing()` | `ui.getSession()` | `stateTree.isPreparingForResync()` |
|---|---|---|---|---|
| resync | `true` | **`false`** | non-null | **`true`** |
| genuine detach | `true` | **`true`** | non-null | `false` |

Note `isAttached()` and `getSession()` are identical in both — during a genuine detach,
`UIInternals#setSession(null)` fires the detach listeners *before* it nulls the session
(`UIInternals.java:517-529`), and `StateNode#setParent` calls `onDetach()` *before* it clears the
parent (`StateNode.java:288-290`). So neither can distinguish the two.

## Consequences

Both reproduced; the second is the more visible, the first the more damaging.

1. **A live tab silently loses its scope.** The resync empties `Lifecycle.uis`, so the scope orphans
   and the scheduled reap destroys it ~60 s later while the tab is happily alive. Afterwards
   `TabScope.getCurrent()` throws *"The TabScope instance is not available for this tab"* and the
   values are gone. There is no recovery path: `prepareForResync` re-fires *attach* listeners, but
   the scope is only ever populated from the `UIInitListener`/ECD callback, which does not re-run.

2. **`IllegalStateException: Invalid state: uis doesn't contain given ui`** when that UI is really
   detached later (while the scope is still alive — e.g. another UI kept it), because `uis` no longer
   holds it. This is the same exception as #5 but a *different* cause, and **the fix chosen for #5
   (stop mutating `uis` in `updateOrphaned()`) does not prevent it** — here `uis` is missing a UI that
   `remove()` legitimately expects.

## Reproduction

Karibu-Testing only, no seams — `prepareForResync()` is the production call site. Both tests assert
the *current* behaviour, so they pass today and should fail once fixed.

```java
public class ResyncIssueTest {
    private static Routes routes;

    @BeforeAll
    public static void createRoutes() {
        routes = new Routes().autoDiscoverViews("com.github.mvysny.vaadin.tabscope");
    }

    @BeforeEach
    public void setupVaadin() {
        MockVaadin.setup(routes);
    }

    @AfterEach
    public void teardownVaadin() {
        try {
            MockVaadin.tearDown();
        } catch (Throwable expected) {
            // the second test leaves the throw above behind
        }
        TabScope.CLEANUP_DURATION_MS = 60 * 1000L;
    }

    @Test
    public void resyncOrphansALiveTabsScope() {
        final String tabA = MockBrowser.getCurrentWindowName();
        final UI ui = UI.getCurrent();
        final TabScope scope = TabScope.getCurrent();
        scope.getValues().setAttribute("precious", "state");
        final AtomicInteger destroyed = new AtomicInteger();
        scope.addDestroyListener(ts -> destroyed.incrementAndGet());

        // exactly what ServerRpcHandler:455 does on a client-requested resynchronization
        ui.getInternals().getStateTree().prepareForResync();

        assertTrue(ui.isAttached(), "the tab is alive: the UI never left the session");
        assertFalse(ui.isClosing(), "...and was never closed");

        TabScope.CLEANUP_DURATION_MS = -1L; // the grace period has 'elapsed'
        MockBrowser.newTab();               // any other tab's init sweeps orphaned scopes
        assertEquals(1, destroyed.get(), "the LIVE tab's scope was destroyed");

        MockBrowser.switchTo(tabA);
        assertThrows(IllegalStateException.class, TabScope::getCurrent);
        assertThrows(IllegalStateException.class, scope::getValues);
    }

    @Test
    public void resyncThenGenuineDetachThrows() {
        final UI ui = UI.getCurrent();
        assertNotNull(TabScope.getCurrent());
        ui.getInternals().getStateTree().prepareForResync();

        // ExecutionException out of Karibu's runUIQueue, cause IllegalStateException
        final ExecutionException e = assertThrows(ExecutionException.class, MockVaadin::tearDown);
        assertEquals("Invalid state: uis doesn't contain given ui", e.getMessage());
    }
}
```

Karibu does not simulate resync itself, so nothing in the current suite covers this path.

## Possible fix

Ignore a detach event for a UI that is not closing:

```java
ui.addDetachListener(e -> {
    if (!ui.isClosing()) {
        return; // resync re-fires detach on a live UI; not a real detach
    }
    removeUI(finalTabScope, ui);
});
```

`isClosing() == true` is a Flow invariant for a genuine detach, not an inference. `VaadinSession#removeUI`
has exactly two callers: `VaadinService#removeClosedUIs` (`:1748`), which only removes UIs where
`isClosing()`, and `VaadinService#fireSessionDestroy` (`:1062`), which closes each UI first and says so
in its own comment:

> `close()` called here for consistency so that it is always called before a UI is removed.
> `UI.isClosing()` is thus always true in `UI.detach()` and associated detach listeners.

Measured with this guard in place: the resync leaves `uis` intact, arms no reap, keeps the values, and
the eventual genuine detach is handled normally — with the whole existing suite still green.

Alternatives considered:

- **`stateTree.isPreparingForResync()`** (`StateTree.java:505`) — the precise discriminator, and true
  exactly while our listener runs, but it lives in `com.vaadin.flow.internal`. The `isClosing()` guard
  gets the same result off public API.
- **Re-add on attach** (`ui.addAttachListener` → `lifecycle.add(ui)`) — `prepareForResync` re-fires
  attach right after the detach, so this self-heals. Weaker on its own, though: it orphans and arms a
  reap before healing, and `add()` rejects a scope that closed in between. Reasonable as belt-and-braces,
  not as the primary fix.

If the guard goes in, the failure direction is the safe one: were Flow ever to add a `removeUI` caller
that doesn't close first, the scope would linger to the session-destroy floor rather than throw.

Found while analysing #5.

**Comment by mvysny** [comment]
Fixed by 0e11237: the UI detach listener acts only when `ui.isClosing()`, so the detach that
`prepareForResync` re-fires on a live UI is ignored and the tab keeps its scope.

Went with `isClosing()` rather than `StateTree#isPreparingForResync()`: the latter is the precise
discriminator but `@since 24.7.5`, and this add-on is `compileOnly` against Vaadin, so consumers on
older versions would get a `NoSuchMethodError` — the same trap as the deprecated-ECD call.
`isClosing()` is established by both callers of `VaadinSession#removeUI`.

`TabScopeResyncTest` covers both directions: a resync costs a live tab nothing (same scope, same
values, no orphaning for the next sweep to reap), and the guard doesn't deafen us to the real detach
— a resynced tab still reaps once it genuinely closes. Reverting the guard turns both red with the
two consequences described above.

**One narrow window is left open on purpose.** Nothing in Flow refuses a request for a closing UI, so
a resync can land between `ui.close()` and the detach that follows at `requestEnd`. There the guard
passes (the UI *is* closing), the UI leaves `uis` early, and its real detach then hits `remove()`'s
`IllegalStateException` — #5's symptom by a much narrower route. Closing it would mean making
`remove()` tolerate an absent UI, which is the fix *not* chosen for #5: it would defang the assertion
and the regression tests that keep `updateOrphaned()` from evicting again. A logged exception at
`requestEnd` costs no tab-scoped state, so that trade stands. INTERNALS.md ("A detach event is not
always a detach") records it, along with the observability table and the `@since` reasoning.
